Add a module parameter to configure the max fetched AWS CFN stack events

denderello opened the pull request:
##### SUMMARY
This change will add a parameter to the `cloudformation` module which allows configuring the maximum number of fetched events from AWS CloudFormation stacks when creating or updating them.

The limit is useful if a stack e.g. exists for a long time or accumulated a lot of changes and therefore iterating over all events using pagination takes too long. Additionally this can be used to prevent running into rate limits when the list of events is very long. 

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
cloudformation

##### ANSIBLE VERSION
```
2.6.0rc2
```

##### ADDITIONAL INFORMATION

denderello committed:
Add a module parameter to configure the max fetched AWS CFN stack events

ryansb commented on lib/ansible/modules/cloud/amazon/cloudformation.py:
```diff
@@ -276,15 +280,16 @@
         ).paginate(
-            StackName=stack_name
+            StackName=stack_name,
+            PaginationConfig={'MaxItems': events_limit}
```
I think it makes sense to only limit the number of events when the `token_filter` isn't passed. That way, if an actual edit was made then all the events resulting from that edit would be returned. CloudFormation has a max number of resource per stack, so there's a natural cap on the number of events per operation. 

What do you think about only having the limit for non-single-op events?

denderello replied:
I digged a bit more in the code and I don't think that this could help us here.

The [DescribeStackEvents API call](https://docs.aws.amazon.com/AWSCloudFormation/latest/APIReference/API_DescribeStackEvents.html) does not allow any kind of server side filter which would allow for limiting the returned events to a specific stack action. Also the Boto Paginator will need to iterate over all pages first before filtering happens client side when using [JMESPath Filtering](https://boto3.readthedocs.io/en/latest/guide/paginators.html#filtering-results-with-jmespath).

Moreover it seems like the `ClientRequestToken` is generated and passed downstream through all function invocations for every action that is not related to CFN Changesets. So I assume that we would actually end up with a set `token_filter` in the `get_stack_events` function every time.

Am I right with my assumptions?

For us it would also be fine having to configure the limit manually and don't provide a default value. We know that our use case is actually a rare one. 😄

ryansb replied:
Yup, that's sensible to have a default limit since we'll be (typically) filtering out most of the historical events with that ClientRequestToken setting.

tedder commented:
Are you using this basically so you can throw away the events? Or what is your use case? I get the rate limit too, so wonder if that's the motivation.

denderello committed:
Fix pep8 errors

denderello committed:
Add version documentation for new configuration option

denderello commented:
@tedder Yes, you are right. We are not interested in the events at all. The only reason why would need them would be for debugging purposes.

Our use case in general is that we are regularly updating stacks as state enforcement on our AWS accounts. Over time (we are operating some stacks which exist for more than 3 years already) the list of events gets quite long and therefore we are not interested in the full history of stack events.

The specific behaviour which we are seeing is that a CFN stack which contains tens of thousands of events will result in a execution time of about 30 minutes for the ansible task. In this situation the CFN stack is updated within a couple of minutes (about 5) while the rest of the time is used for fetching all events.

denderello committed:
Mark module option for ansible version 2.7

denderello committed:
Adapt tests to cloudformation module changes

ryansb commented:
Thanks for the insight into your use case. 

It seems like the resource limit is 200 per stack (I'm not sure if there are places that can raise that), so maybe we could bump the default to 200 so each resource could (in theory) produce one event. The DescribeStackEvents API seems to return 100 at a time, so that would only require two calls to fill in. 

If you'd rather leave it at 100 that's ok, but I feel like matching the resource limit would be pretty sensible.

ryansb reviewed (approved):
Looks good overall

denderello committed:
Increase default events limit in cloudformation module

In order to make sure that enough are fetched by default. This align roughly with the limit if manageable resources in CloudFormation.

denderello commented:
200 is totally fine for us as well and with the variable being configurable it should be no problem to change this. I already pushed the change of the default. 

Thanks a lot for the input and review!

ryansb commented:
This works for me. Tested with cutting events down to just 2, setting a limit over the number of events I have on the stack, and setting the limit to 0. Great!
